structures/segment-tree/problems/coming-and-going/coming-and-going.cpp

```diff
@@ -0,0 +1,199 @@
+	update_segment_tree(tree, start, mid, changed_idx, new_val, (2 * st_idx) + 1);
+	update_segment_tree(tree, mid + 1, end, changed_idx, new_val, (2 * st_idx) + 2);
+
+	return;
+}
+
+// #ifdef hackpackpp
+// find the largest value for a range
+//
+// tree        - the segment tree to query
+// start       - the start index of the _source_ array
+// end         - the end index of the _source_ array
+// range_start - the start index of the query
+// range_end   - the end index of the query
+// greatest    - the greatest value found so far (set to zero when first calling)
+// st_idx      - the current index in the segtree (set to zero when first calling)
+// #endif
+int query_segment_tree(const int* const tree, const unsigned int start, const unsigned int end, const unsigned int range_start, const unsigned int range_end, int greatest, const unsigned int st_idx)
+{
+	// #ifdef hackpackpp
+	// out of range; do not continue
+	// #endif
+	if(start > range_end || end < range_start)
+		return -2;
+
+	// #ifdef hackpackpp
+	// leaf node; the greatest value one can see from here
+	// is the value in the leaf node
+	// #endif
+	if(start == end)
+		return tree[st_idx];
+
+	// #ifdef hackpackpp
+	// this node can be considered if the values in the original
+	// array it represents fit in the query's range
+	// #endif
+	if(start >= range_start && end <= range_end)
+	{
+		if(tree[st_idx] > greatest)
+			return tree[st_idx];
+	}
+	else
+	{
+		const unsigned int mid = start + ((end - start) / 2);
+		const int left = query_segment_tree(tree, start, mid, range_start, range_end, greatest, (2 * st_idx) + 1);
+		const int right = query_segment_tree(tree, mid + 1, end, range_start, range_end, greatest, (2 * st_idx) + 2);
+
+		// #ifdef hackpackpp
+		// use the greater of the two values here;
+		// if a summation were desired, one could simply add the
+		// two values together
+		// #endif
+		if(left > greatest)
+			greatest = left;
+		if(right > greatest)
+			greatest = right;
+	}
+
+	return greatest;
+}
+
+int main()
+{
+	
+	int headcount[N] = { 0 };
+
+	// #ifdef hackpackpp
+	// size needed for segtree is 2^{ceil(log2(n)) + 1} - 1
+	// #endif
+	const unsigned int st_size = pow(2, ceil(log2(N)) + 1) - 1;
+	int* segtree = new int[st_size];
+	build_segment_tree(headcount, 0, N - 1, segtree, 0);
+	
+	unsigned int no_cows;
+	cin >> no_cows;
+
+	for(unsigned int i = 0; i < no_cows; i++)
+	{
+		int t1, t2;
+		unsigned int in_barn_times;
+		cin >> t1 >> t2 >> in_barn_times;
+
+		for(unsigned int j = 0; j < in_barn_times; j++)
+		{
+			unsigned int start, end;
+			cin >> start >> end;
+			for(unsigned int k = start; k <= end; k++)
+			{
+				// #ifdef hackpackpp
+				// account for times 0 to 5 we don't care about with k - 6
+				// #endif
+				headcount[k - 6]++;
+				update_segment_tree(segtree, 0, N - 1, k - 6, headcount[k - 6], 0);
+			}
+		}
+
+		// #ifdef hackpackpp
+		// print out cow count if requested as data comes in
+		// #endif
+		if(t1 && t2)
```

Comment:
even though `(t1 && t2)` is shorter than `(t1 == 0 && t2 == 0)` it is a bit more readable.